tests/python/nightly/frontend/nnef/test_forward.py

```diff
@@ -0,0 +1,1629 @@
+
+import tvm
+import tvm.testing
+from tvm import relay
+
+import cases_string
+
+
+def get_case_graph(name):
+    if "-" in name:
+        name = name.replace("-", "_")
+    return nnef.parse_string(getattr(cases_string, name))
+
+
+def get_nnef_outputs(path, inputs):
+    ip = interpreter.Interpreter(path, None, None)
+    inputs = [inputs[tensor.name] for tensor in ip.input_details()]
+    return ip(inputs)
+
+
+def get_type(val):
+    if val == "scalar":
+        return "float32"
+    if val == "integer":
+        return "int32"
+    if val == "logical":
+        return "bool"
+    if val == "string":
+        return "string"
+
+
+def verify_model(
+    model_path,
+    target,
+    device,
+    rtol=1e-5,
+    atol=1e-5,
+):
+    graph = get_case_graph(model_path)
+    nnef.infer_shapes(graph)
+    inputs = {}
+    # generate inputs
+    for inp in graph.inputs:
+        intensor = graph.tensors[inp]
+        shape = intensor.shape
+        if any(exc in model_path for exc in ["log", "sqrt", "pow", "batch_norm"]):
+            low = 0.0
+        else:
+            low = -1.0
+        high = 1.0
+        if "acosh" in model_path:
+            high = 2.0
+            low = 1.0
+        if intensor.dtype == "scalar":
+            inputs[inp] = np.random.uniform(low=low, high=high, size=shape).astype("float32")
+        elif intensor.dtype == "integer":
+            inputs[inp] = np.random.randint(0, 64, shape)
+        elif intensor.dtype == "logical":
+            inputs[inp] = np.random.binomial(1, 0.5, shape).astype("bool")
+        elif intensor.dtype == "string":
+            inputs[inp] = np.random.uniform(low=low, high=high, size=shape).astype("string")
+
+    # set graph parameters
+    for operation in graph.operations:
+        if operation.name == "variable":
+            tensor_name = operation.outputs["output"]
+
+            shape = operation.attribs["shape"]
+
+            assert (
+                operation.dtype == "scalar"
+            ), f"variable of type {operation.dtype} is not supported, please update verify_model"
+
+            data = np.random.uniform(low=-1.0, size=shape).astype("float32")
+
+            tensor = graph.tensors[tensor_name]
+            graph.tensors[tensor_name] = _nnef.Tensor(
+                tensor.name, tensor.dtype, shape, data, tensor.quantization
+            )
+
+    outputs = get_nnef_outputs(graph, inputs)
+
+    mod, params = relay.frontend.from_nnef(graph)
+
+    with tvm.transform.PassContext(opt_level=3):
+        # dev = tvm.device(target, 0)
+        executor = relay.create_executor(
+            "graph", mod, device=device, target=target, params=params
+        ).evaluate()
+        out = executor(**inputs)
+
+        if not isinstance(out, (list, tuple)):
+            out = [out]
+
+        for i, base_out in enumerate(outputs):
+            tvm.testing.assert_allclose(out[i].numpy(), outputs[base_out], rtol=rtol, atol=atol)
+
+
+# graph tests
+
+
+@tvm.testing.parametrize_targets
```

Comment:
it would be good to add `dim` as an argument to veryfy_model, and we can use @pytest.mark.parametrize to feed various dim